go-controller/pkg/ovn/controller/services/services_controller_test.go

```diff
@@ -440,6 +518,13 @@ func TestSyncServices(t *testing.T) {
 			}
 
 			g.Eventually(controller.nbClient).Should(libovsdbtest.HaveData(tt.expectedDb))
+
+			if tt.nodeToDelete != nil {
+				controller.nodeTracker.removeNode(tt.nodeToDelete.name)
+				g.Expect(controller.syncService(namespacedServiceName(ns, serviceName))).To(gomega.Succeed())
+				g.Eventually(controller.nbClient).ShouldNot(libovsdbtest.ContainData(nodeRouterLoadBalancer(firstNode, nodePort, serviceName, ns, outport, nodeAEndpointIP, nodeBEndpointIP)))
+				g.Eventually(controller.nbClient).Should(libovsdbtest.HaveDataIgnoringUUIDs(tt.dbStateAfterDeleting))
```

Comment:
It doesn't work, with the following mismatch:

```
the missing elements were
            <[]*libovsdb.testDataMatcher | len:2, cap:2>: [
                {
                    expected: <*nbdb.LogicalSwitch | 0xc00149e1c0>{
                        UUID: "4d35a473-6481-4803-89a6-cb9026decb8a [switch-node-a]",
                        ACLs: nil,
                        Copp: nil,
                        DNSRecords: nil,
                        ExternalIDs: nil,
                        ForwardingGroups: nil,
                        LoadBalancer: [
                            "Service_testns/foo_TCP_cluster",
                        ],
                        LoadBalancerGroup: nil,
                        Name: "switch-node-a",
                        OtherConfig: nil,
                        Ports: nil,
                        QOSRules: nil,
                    },
                    ignoreUUID: false,
                },
                {
                    expected: <*nbdb.LogicalRouter | 0xc000146750>{
                        UUID: "700c91c2-0948-45fe-9aed-e8fa0f81227b [gr-node-a]",
                        Copp: nil,
                        Enabled: nil,
                        ExternalIDs: nil,
                        LoadBalancer: [
                            "Service_testns/foo_TCP_cluster",
                        ],
                        LoadBalancerGroup: nil,
                        Name: "gr-node-a",
                        Nat: nil,
                        Options: nil,
                        Policies: nil,
                        Ports: nil,
                        StaticRoutes: nil,
                    },
                    ignoreUUID: false,
                },
            ]
        the extra elements were
            <[]interface {} | len:2, cap:2>: [
                <*nbdb.LogicalRouter | 0xc000146b60>{
                    UUID: "700c91c2-0948-45fe-9aed-e8fa0f81227b [gr-node-a]",
                    Copp: nil,
                    Enabled: nil,
                    ExternalIDs: nil,
                    LoadBalancer: [
                        "eb32f513-9e84-49e3-8975-ad645116de1e",
                    ],
                    LoadBalancerGroup: nil,
                    Name: "gr-node-a",
                    Nat: nil,
                    Options: nil,
                    Policies: nil,
                    Ports: nil,
                    StaticRoutes: nil,
                },
                <*nbdb.LogicalSwitch | 0xc00149e620>{
                    UUID: "4d35a473-6481-4803-89a6-cb9026decb8a [switch-node-a]",
                    ACLs: nil,
                    Copp: nil,
                    DNSRecords: nil,
                    ExternalIDs: nil,
                    ForwardingGroups: nil,
                    LoadBalancer: [
                        "eb32f513-9e84-49e3-8975-ad645116de1e",
                    ],
                    LoadBalancerGroup: nil,
                    Name: "switch-node-a",
                    OtherConfig: nil,
                    Ports: nil,
                    QOSRules: nil,
                },
            ]
```

I.e. the associated load-balancers on the re-added node / router feature UUIDs, while we expect those to be names. 

If you suggest an alternative, I can try it out @jcaamano .

Reply:
Somehow, seems the full re-sync put the UUIDs there. Does this make sense ?

Reply:
Expected LB (consistOf stanza - I get really confused about gingko sometimes ...):
```
{
    UUID: "c827591f-14d2-4b97-b015-9fdee6d81202 [Service_testns/foo_TCP_cluster]",
    ExternalIDs: {
        "k8s.ovn.org/kind": "Service",
        "k8s.ovn.org/owner": "testns/foo",
    },
    HealthCheck: nil,
    IPPortMappings: nil,
    Name: "Service_testns/foo_TCP_cluster",
    Options: {"skip_snat": "false", "event": "false", "reject": "true"},
    Protocol: "tcp",
    SelectionFields: nil,
    Vips: {
        "192.168.1.1:80": "10.128.0.2:3456,10.128.1.2:3456",
    },
},
```

Actual:
```
<*nbdb.LoadBalancer | 0xc00002b080>{
                    UUID: "c827591f-14d2-4b97-b015-9fdee6d81202 [Service_testns/foo_TCP_cluster]",
                    ExternalIDs: {
                        "k8s.ovn.org/owner": "testns/foo",
                        "k8s.ovn.org/kind": "Service",
                    },
                    HealthCheck: nil,
                    IPPortMappings: nil,
                    Name: "Service_testns/foo_TCP_cluster",
                    Options: {"reject": "true", "skip_snat": "false", "event": "false"},
                    Protocol: "tcp",
                    SelectionFields: nil,
                    Vips: {
                        "192.168.1.1:80": "10.128.0.2:3456,10.128.1.2:3456",
                    },
                },
```

The load balancers themselves **are the same** - otherwise they would be listed in the differences. 

The routers / switches pointing at them just feature names / UUIDs, which, as you say, is not the same. 

I don't know a way for this to work without using the `HaveDataIgnoringUUIDs` matcher. All ears though.

@jcaamano

Reply:
Something is wrong with the test or the code under test because
 `eb32f513-9e84-49e3-8975-ad645116de1e != c827591f-14d2-4b97-b015-9fdee6d81202`

Can you show the contents of the `eb32f513-9e84-49e3-8975-ad645116de1e`LB in actual.

Reply:
OK I see now what is happening, thanks for making me look deeper onto what is happening here.

Somehow, when the service is reconciled, the load balancers that are being removed from the deleted node logical switch / gateway router are the cluster-wide ones - `Service_testns/foo_TCP_cluster`. 

I somehow expected the load balancers that were dissociated from the logical switch / gw router to be the ones related to the node -  `Service_testns/foo_TCP_node_router+switch_node-a`. 
Thinking on this a bit more ... I think what really **should** happen is for **all** load balancers to be removed from the node's logical switch / gw router when the node is deleted.

I think this "doesn't matter" in the big picture, because the node's logical switches / gateway routers (which feature incorrect load balancers) will be deleted - as a result of the k8s node being deleted. Thus, having them feature the incorrect association does not matter much, since all pods are being evicted.

... never the less, this is a unit test. The service reconciler should work correctly no matter if the logical sw / gw router are deleted afterwards. I'll keep looking.

Reply:
Cleaning up the extra load balancer would require adding code to the production code to account for unit test behavior - the real OVSDB server would remove these LB entries from the logical switch / gw router (since they are weak references).

I do not think that is OK. 

I would rather push for allowing the `HaveDataIgnoringUUIDs` matcher, given the reasons explained above. 

Do you see an alternative @jcaamano ?

Reply:
I would get get the real UUIDs of the LBs that you expect to be removed from the DB before actually removing the node, and set them in the expected LS and LR at the end. I would still not use `HaveDataIgnoringUUIDs`.

Reply:
Would you take another look @jcaamano ? 

I think this covers it.

This "hack" could be removed if we somehow add weak reference cleanup to the test framework.

I could try to investigate that, and eventually (... hopefully) propose a follow-up PR for that.